### PR TITLE
.*: add list-member command

### PR DESCRIPTION
**Pull request by GMHDBJD**
### What problem does this PR solve? 
- add list-member command in DM-HA

### What is changed and how it works?
some interface preview
Screenshot (19)
Screenshot (22)
Screenshot (20)
Screenshot (21)

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change
 - Has interface methods change

### CODE REVIEW

File: dm/master/server.go [hunk 1]
```diff
@@ -1559,3 +1559,24 @@ func (s *Server) getSourceRespsAfterOperation(ctx context.Context, taskName stri
 	wg.Wait()
 	return sortCommonWorkerResults(sourceRespCh)
 }
+
+// GetLeader get master leader
+func (s *Server) GetLeader(ctx context.Context, req *pb.GetLeaderRequest) (*pb.GetLeaderResponse, error) {
```
Comment (WangXiangUSTC) [inline dm/master/server.go:1564]: do we need to forward this request to the leader just like other requests? @csuzhangxc

File: dm/master/server_test.go [hunk 2]
```diff
@@ -1141,3 +1141,86 @@ func mockRevelantWorkerClient(mockWorkerClient *pbmock.MockWorkerClient, taskNam
...
+		res, err := s2.GetLeader(ctx, &pb.GetLeaderRequest{})
+		c.Assert(err, check.IsNil)
+		c.Assert(res.Name, check.Equals, "")
+		c.Assert(res.Addr, check.Equals, "")
+	}
+
+	/*
+		It seems that if we kill leader in 2 masters cluster,
+		the second one wouldn't be the leader so follow cases would not be passed.
+		But I fail to create 3 masters cluster...
```
Comment (WangXiangUSTC) [inline dm/master/server_test.go:1211]: what is the error log of creating the third master?

Comment (GMHDBJD) [reply dm/master/server_test.go:1211]: The problem has been solved by waiting for the second master start.

File: dm/ctl/master/list_member.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+	"github.com/pingcap/dm/dm/pb"
+
+	"github.com/pingcap/errors"
+	"github.com/spf13/cobra"
+)
+
+// NewListMemberCmd creates an ListMember command
+func NewListMemberCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "list-member master/worker/leader",
```
Comment (csuzhangxc) [inline dm/ctl/master/list_member.go:31]: Can we list all members if no more argument specified? in other words, make `master`/`worker`/`leader` become optional arg (`[--master] [--worker] [--leader]`).

File: dm/proto/dmmaster.proto [hunk 4]
```diff
@@ -62,6 +62,22 @@ service Master {
...
     rpc OfflineWorker(OfflineWorkerRequest) returns(OfflineWorkerResponse) {}
+
+    // ListMember list member master information
+    rpc ListMemberMaster(ListMemberRequest) returns(ListMemberMasterResponse) {}
+
+    // ListMember list member worker information
+    rpc ListMemberWorker(ListMemberRequest) returns(ListMemberWorkerResponse) {}
+
+    // ListMember list member leader information
+    rpc ListMemberLeader(ListMemberRequest) returns(ListMemberLeaderResponse) {}
```
Comment (csuzhangxc) [inline dm/proto/dmmaster.proto:73]: Can we combine them into one method?

Comment (GMHDBJD) [reply dm/proto/dmmaster.proto:73]: But their return type are different.

Comment (csuzhangxc) [reply dm/proto/dmmaster.proto:73]: Is `oneof` can work here? (as we did in `message SubTaskStatus`).

Comment (GMHDBJD) [reply dm/proto/dmmaster.proto:73]: Got it. I will have a try.

File: dm/ctl/master/list_member.go [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+import (
+	"context"
+	"os"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/pingcap/dm/dm/ctl/common"
+	"github.com/pingcap/dm/dm/pb"
+
+	"github.com/pingcap/errors"
+	"github.com/spf13/cobra"
```
Comment (WangXiangUSTC) [inline dm/ctl/master/list_member.go:25]: need to format these pkg

File: dm/master/server.go [hunk 6]
```diff
@@ -1559,3 +1559,130 @@ func (s *Server) getSourceRespsAfterOperation(ctx context.Context, taskName stri
...
+		return nil, terror.ErrMasterRequestIsNotForwardToLeader
+	}
+
+	memberList, err := s.etcdClient.MemberList(ctx)
+	if err != nil {
+		resp.Msg = errors.ErrorStack(err)
+		return resp, nil
+	}
+
+	_, leaderName, _, err := s.election.LeaderInfo(ctx)
```
Comment (WangXiangUSTC) [inline dm/master/server.go:1658]: seems that the information returned by `LeaderInfo` is enough, don't need to query etcd again? and we can update the result of `LeaderInfo` if necessary

**Comment by WangXiangUSTC** [comment]
rest LGTM

**Comment by GMHDBJD** [comment]
@WangXiangUSTC @csuzhangxc @lichunzhu all comments have been addressed. PTAL

File: dm/master/server.go [hunk 7]
```diff
@@ -1559,3 +1559,141 @@ func (s *Server) getSourceRespsAfterOperation(ctx context.Context, taskName stri
...
+		resp.Leader.Msg = errors.ErrorStack(err)
+		return resp, nil
+	}
+
+	client := http.Client{
+		Timeout: 1 * time.Second,
+	}
+
+	alive := true
+	_, err = client.Get("http://" + addr + "/health")
```
Comment (WangXiangUSTC) [inline dm/master/server.go:1637]: in which case the leader is not alive?

Comment (GMHDBJD) [reply dm/master/server.go:1637]: Maybe when we enter this function the leader change and immediately down? Alright I will remove it. 😂

File: dm/ctl/master/list_member.go [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+)
+
+// NewListMemberCmd creates an ListMember command
+func NewListMemberCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "list-member [--master] [--worker] [--leader]",
+		Short: "list member information",
+		Run:   listMemberFunc,
+	}
+	cmd.Flags().BoolP("master", "", false, "only to list master")
```
Comment (lichunzhu) [inline dm/ctl/master/list_member.go:34]: I think we can use `StringSliceVarP` here to query specified master/worker status.

Comment (GMHDBJD) [reply dm/ctl/master/list_member.go:34]: But `list-member --master` will fail. Can we use `list-member --master=*` instead?

File: dm/ctl/master/list_member.go [hunk 9]
```diff
@@ -0,0 +1,98 @@
...
+
+// Reset clears cache of ListMemberFlags
+func (c ListMemberFlags) Reset() {
+	c.names = c.names[:0]
+}
+
+// NewListMemberCmd creates an ListMember command
+func NewListMemberCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "list-member [--type] [--name]",
```
Comment (lichunzhu) [inline dm/ctl/master/list_member.go:44]:
```suggestion
		Use:   "list-member [--type leader/master/worker] [--name master-name/worker-name ...]",
```

File: tests/ha_master/run.sh [hunk 10]
```diff
@@ -12,6 +12,171 @@ MASTER_PORT3=8461
...
+            exit 1
+        fi
+    done
+    run_dm_ctl_with_retry $WORK_DIR "127.0.0.1:$MASTER_PORT" \
+        "list-member --type=master" \
+        "\"alive\": true" 4
+    echo "kill leader" $leader
+    ps aux | grep $leader |awk '{print $2}'|xargs kill || true
+    check_port_offline ${master_ports[$leader_idx]} 20
+    sleep 5
```
Comment (lichunzhu) [inline tests/ha_master/run.sh:71]: It sames that these code are exactly the same with L73-94. Can we put them in one for-loop?

File: dm/master/scheduler/scheduler.go [hunk 11]
```diff
@@ -573,6 +573,22 @@ func (s *Scheduler) RemoveWorker(name string) error {
 	return nil
 }
 
+// GetWorker gets all worker agent.
+func (s *Scheduler) GetWorker() ([]*Worker, error) {
```
Comment (lichunzhu) [inline dm/master/scheduler/scheduler.go:577]:
```suggestion
func (s *Scheduler) GetAllWorkers() ([]*Worker, error) {
```

File: dm/ctl/master/list_member.go [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+}
+
+// NewListMemberCmd creates an ListMember command
+func NewListMemberCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "list-member [--type leader/master/worker] [--name master-name/worker-name ...]",
+		Short: "list member information",
+		Run:   listMemberFunc,
+	}
+	cmd.Flags().StringP("type", "t", "", "specify a member type from master, worker, and leader")
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply dm/ctl/master/list_member.go:48]: Actually comma is needed here. See https://github.com/pingcap/docs-dm/pull/66#discussion_r422449521

[the remark this answers is not in the document]
Comment (lichunzhu) [reply dm/ctl/master/list_member.go:48]: alright

**Comment by GMHDBJD** [comment]
PTAL. @WangXiangUSTC @csuzhangxc @lichunzhu

File: dm/ctl/master/list_member.go [hunk 13]
```diff
@@ -0,0 +1,105 @@
...
+
+// ListMemberFlags are flags that used in ListMember command
+type ListMemberFlags struct {
+	names []string // specify names to list information
+}
+
+// Reset clears cache of ListMemberFlags
+func (c ListMemberFlags) Reset() {
+	c.names = c.names[:0]
+}
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply dm/ctl/master/list_member.go]: Not necessary. I have removed it.

**Pull request merged**